dbms/src/Storages/Page/V3/Universal/UniversalPageStorage.cpp

```diff
@@ -461,4 +479,59 @@ void UniversalPageStorage::dumpIncrementalCheckpoint(const UniversalPageStorage:
     last_checkpoint_sequence = snap->sequence;
 }
 
+std::unordered_set<String> UniversalPageStorage::getRemoteFileIdsNeedCompact(const double gc_threshold)
+{
+    // In order to not block local GC updating the cache, get a copy of the cache
+    auto stats = remote_data_files_stat_cache.getCopy();
+
+    auto s3_client = S3::ClientFactory::instance().sharedTiFlashClient();
+
+    // If the total_size is 0, try to get the actual size from S3
+    std::vector<std::future<std::tuple<String, size_t>>> actual_sizes;
+    for (const auto & [file_id, stat] : stats)
+    {
+        if (stat.total_size != 0)
+        {
+            continue;
+        }
+
+        auto task = std::make_shared<std::packaged_task<std::tuple<String, size_t>()>>(
+            [&s3_client, file_id = file_id, log = this->log]() noexcept {
+                auto view = S3::S3FilenameView::fromKey(file_id);
+                auto datafile_key = view.asDataFile().toFullKey();
+                auto actual_total_size = S3::getObjectSize(*s3_client, s3_client->bucket(), datafile_key, "", /*throw_on_error*/ false);
+                if (actual_total_size == 0)
+                {
+                    LOG_WARNING(log, "failed to get S3 object size, file_id={}", file_id);
+                }
+                return std::make_tuple(file_id, actual_total_size);
+            });
+        actual_sizes.emplace_back(task->get_future());
+        IOThreadPool::get().scheduleOrThrowOnError([task] { (*task)(); });
+    }
+    for (auto & f : actual_sizes)
+    {
+        const auto & [file_id, actual_size] = f.get();
+        auto iter = stats.find(file_id);
+        RUNTIME_CHECK(iter != stats.end(), file_id);
+        iter->second.total_size = actual_size;
+    }
+
+    // update cache by the S3 result
+    remote_data_files_stat_cache.updateTotalSize(stats);
+
+    std::unordered_set<String> rewrite_files;
+    for (const auto & [file_id, stat] : stats)
+    {
+        if (stat.total_size == 0)
+            continue;
+        double valid_rate = 1.0 * stat.valid_size / stat.total_size;
+        if (valid_rate < gc_threshold)
+        {
+            rewrite_files.emplace(file_id);
+        }
+    }
+    return rewrite_files;
```

Comment:
TODO: when the file size is less than 128KiB, then rewrite it
https://aws.amazon.com/cn/s3/pricing/